storage/innobase/srv/srv0srv.cc

```diff
@@ -811,11 +811,10 @@ srv_printf_innodb_monitor(
 	for (ulint i = 0; i < btr_ahi_parts && btr_search_enabled; ++i) {
 		const auto part= &btr_search_sys.parts[i];
 		part->latch.rd_lock(SRW_LOCK_CALL);
-		ut_ad(part->heap->type == MEM_HEAP_FOR_BTR_SEARCH);
 		fprintf(file, "Hash table size " ULINTPF
 			", node heap has " ULINTPF " buffer(s)\n",
 			part->table.n_cells,
-			part->heap->base.count - !part->heap->free_block);
+			part->heap->base.count - !part->heap->ahi_block);
 		part->latch.rd_unlock();
```

Comment:
If `btr_search_enabled` had been cleared between our check and the acquisition of `part->latch`, we’d be dereferencing `part->heap=nullptr` here.

A possible fix could be to ensure that `part->heap` will always remain allocated. In that way, we would not have to check for `btr_search_enabled` here at all, and `btr_search_check_free_space_in_heap()` could safely use atomic memory access instead of acquiring `part->latch`.